servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/NewRoundRobinLoadBalancer.java

```diff
@@ -250,59 +251,88 @@ public void onNext(@Nullable final Collection<? extends ServiceDiscovererEvent<R
-                                    return emptyList();
-                                } else {
-                                    return markHostAsExpired(oldHostsTyped, addr);
-                                }
-                            } else if (UNAVAILABLE.equals(eventStatus)) {
-                                return listWithHostRemoved(oldHostsTyped, host -> {
-                                    boolean match = host.address().equals(addr);
-                                    if (match) {
-                                        host.markClosed();
-                                    }
-                                    return match;
-                                });
-                            } else {
-                                LOGGER.error("{}: Unexpected Status in event:" +
-                                        " {} (mapped to {}). Leaving usedHosts unchanged: {}",
-                                        NewRoundRobinLoadBalancer.this, event, eventStatus, oldHosts);
-                                return oldHosts;
-                            }
-                        });
 
-                LOGGER.debug("{}: now using addresses (size={}): {}.",
-                        NewRoundRobinLoadBalancer.this, usedAddresses.size(), usedAddresses);
+            boolean sendReadyEvent;
+            List<Host<ResolvedAddress, C>> nextHosts = new ArrayList<>();
+            for (;;) {
+                // TODO: we have some weirdness in the event that we fail the CAS namely that we can create a host
+                //  that never gets used but is orphaned. It's fine so long as there is nothing to close but that
+                //  guarantee may not always hold in the future.
+                List<Host<ResolvedAddress, C>> usedHosts = usedHostsUpdater.get(NewRoundRobinLoadBalancer.this);
+                if (isClosedList(usedHosts)) {
+                    // We don't update if the load balancer is closed.
+                    return;
+                }
+                nextHosts.clear();
+                sendReadyEvent = false;
+
+                // First we make a map of addresses to events so that we don't get quadratic behavior for diffing.
+                // Unfortunately we need to make this every iteration of the CAS loop since we remove entries
+                // for hosts that already exist. If this results in to many collisions and map rebuilds we should
+                // re-assess how we manage concurrency for list mutations.
+                final Map<ResolvedAddress, ServiceDiscovererEvent<ResolvedAddress>> eventMap = new HashMap<>();
+                for (ServiceDiscovererEvent<ResolvedAddress> event : events) {
+                    ServiceDiscovererEvent<ResolvedAddress> old = eventMap.put(event.address(), event);
+                    if (old != null) {
+                        LOGGER.error("Multiple ServiceDiscoveryEvent's detected for address {}. Event: {}.",
```

Comment:
According to the current agreement, this is not an error. It's ok to receive multiple events for the same address. LB impl can choose how to treat that behavior:
1. RR always takes the latest one.
2. If LB considers meta-data, it should either take the latest mata-data or combine all meta-data between events together.
3. Some other LB may consider multiple appearance of the same address as an increased priority for that address.

If you would like to say that this particular LB doesn't like that behavior, consider using `debug` or `info` level. `error` will scary users bcz they have queries to monitor any errors.

Reply:
I switched it to behave same as RR and log at debug.
As an aside, the lack of contract here is kind of difficult to program around and blurs the separation of SD and LB a good amount since they can start to be behavior dependent, not just API dependent. What are the odds we could/want to tighten up the contract?

Reply:
IIRC, the first problem was that it was hard to make sure all SD implementations behave identically (I know a few). It's not trivial to explain users how to program it correctly and filter out duplicates, taking into account retries and preserving a state between retries in reactive streams. So, we made the contract easier and said that it's simpler to take care of it at LB layer. And in addition, we considered use-cases described in my first message when LB can have flexibility how to interpret events.

We can absolutely tighten them up for cases when we support both implementations, but it's hard to tighten something that we do not control.